example/offload-template/src/StepDiagnosticData.hh

```diff
@@ -0,0 +1,157 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file offload-template/src/StepDiagnosticData.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <celeritas/Quantities.hh>
+#include <celeritas/Types.hh>
+#include <celeritas/Units.hh>
+#include <corecel/Macros.hh>
+#include <corecel/data/Collection.hh>
+#include <corecel/data/CollectionAlgorithms.hh>
+#include <corecel/data/CollectionBuilder.hh>
+#include <corecel/data/Filler.hh>
+
+namespace celeritas
+{
+namespace example
+{
+//---------------------------------------------------------------------------//
+//! Step statistics gathered inside a kernel
+struct NativeStepStatistics
+{
+    using real_type = double;
+
+    real_type step_length{};
+    real_type energy_deposition{};  // MeV
+};
+}  // namespace example
+
+#if CELER_USE_DEVICE
+// The Filler is instantiated in StepDiagnosticData.cu
+extern template class Filler<example::NativeStepStatistics, MemSpace::device>;
+#endif
+
+namespace example
+{
+//---------------------------------------------------------------------------//
+//! Step statistics gathered in host memory
+struct HostStepStatistics
+{
+    //! Accumulated number of steps
+    size_type steps{};
+
+    //! Accumulated number of new tracks
+    size_type generated{};
+
+    //! Accumulated number of secondaries
+    size_type secondaries{};
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Shared setup data.
+ *
+ * This class is where problem setup data, including variable-length data that
+ * gets copied to device, is defined.
+ */
+template<Ownership W, MemSpace M>
+struct StepParamsData
+{
+    static constexpr EventId::size_type num_events{1};
+
+    //! Whether the class is non-empty
+    explicit CELER_FUNCTION operator bool() const { return true; }
+
+    //! Assign from another set of data (null-op since no member data)
+    template<Ownership W2, MemSpace M2>
+    StepParamsData& operator=(StepParamsData<W2, M2> const&)
+    {
+        return *this;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Manage data ownership and reference and transfer between host/device.
+ *
+ * Only some of the data (\c NativeStepStatistics) is updated on device. The
+ * number of steps must be changed outside the kernel.
+ */
+template<Ownership W, MemSpace M>
+struct StepStateData
+{
+    //// TYPES ////
+
+    using size_type = unsigned long long;
```

Comment:
Is this used anywhere? Should it be in the `HostStepStatistics`?

Reply:
Yep that's where it was meant to be!